OCPBUGS-27376: Allow domain names starting with a number

CrystalChun opened the pull request:
[OCPBUGS-27376](https://issues.redhat.com/browse/OCPBUGS-27376)
[RFC 1034](https://datatracker.ietf.org/doc/html/rfc1034#section-3.5) initially stated domain names should only start with a letter, but that was later revised by [RFC 1123](https://datatracker.ietf.org/doc/html/rfc1123#page-13). Updates the domain regex to allow domains to start with a number. Also adds a check to prevent dotted decimal domains. 

Includes changes that also check for the length of domains. [RFC 1123](https://datatracker.ietf.org/doc/html/rfc1123#page-13) also states that a label must be no more than 63 characters and a full domain must be no more than 255 characters.

Revision to the previous incorrect PR https://github.com/openshift/assisted-service/pull/5801
<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

You can refer to [Kubernetes community documentation] on writing good commit messages, which provides good tips and ideas.

Some PRs address specific issues. Please, refer to the [CONTRIBUTING] documentation for more
information on how to link a PR to an existing issue.

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer?
- Is this PR relying on CI for an e2e test run?
- Should this PR be tested in a specific environment?
- Any logs, screenshots, etc that can help with the review process?

-->

## List all the issues related to this PR

- [ ] New Feature <!-- new functionality -->
- [ ] Enhancement <!-- refactor, code changes, improvement, that won't add new features -->
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD <!-- Notice that changes for Dockerfiles/Jenkinsfiles aren't tested in CI due to a known bug. -->

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

<!-- Please, select one or more if needed: -->

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

CrystalChun commented:
/test ?

CrystalChun commented:
/test edge-unit-test edge-subsystem-aws

CrystalChun commented:
/retest-required

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_assisted-service/5914/pull-ci-openshift-assisted-service-master-edge-subsystem-kubeapi-aws/1749943617230737408

Looks like a flaky test 
```
{ Failure /assisted-service/subsystem/kubeapi_test.go:2574
Expected
    <*errors.StatusError | 0xc001382960>: 
    Operation cannot be fulfilled on infraenvs.agent-install.openshift.io "infraenv-1fb1886b": the object has been modified; please apply your changes to the latest version and try again
    {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "Operation cannot be fulfilled on infraenvs.agent-install.openshift.io \"infraenv-1fb1886b\": the object has been modified; please apply your changes to the latest version and try again",
            Reason: "Conflict",
            Details: {
                Name: "infraenv-1fb1886b",
                Group: "agent-install.openshift.io",
                Kind: "infraenvs",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 409,
        },
    }
to be nil
/assisted-service/subsystem/kubeapi_test.go:2592}
```

carbonin commented on pkg/validations/validations.go:
```diff
@@ -197,3 +200,11 @@ func ValidateCaCertificate(certificate string) error {
+func isDottedDecimalDomain(domain string) bool {
+	regex := `([\d]+\.){3}[\d]+`
+	return regexp.MustCompile(regex).MatchString(domain)
+
```
Bonus line not needed.

CrystalChun replied:
Thanks for catching that! Removed the extra line

carbonin reviewed (commented):
Is there a test for the length requirement?

It's not clear that the `hugeDomainName` would be valid if not for length.

It might be worth adding a descriptor to each test case explaining why the name fails.

CrystalChun commented:
> Is there a test for the length requirement?

> It's not clear that the hugeDomainName would be valid if not for length.

> It might be worth adding a descriptor to each test case explaining why the name fails.

Thanks for your review! I added a "reason" field to the unit tests and modified the second test using `fqnHugeDomain` to have > 255 characters. What do you think?

CrystalChun commented:
/jira-refresh

CrystalChun commented:
/jira refresh

CrystalChun commented:
/cherry-pick cloud_hotfix_releases

CrystalChun commented:
/cherry-pick release-ocm-2.9

CrystalChun commented:
/cherry-pick release-ocm-2.8

CrystalChun commented:
/cherry-pick release-ocm-2.7
